x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,453 @@
+	require.NoError(t, err)
+
+	setup.UpgradeKeeper.AssertCalled(
+		t,
+		"ScheduleUpgrade",
+		mock.MatchedBy(isContextWithNonCachedMultiStore),
+		msgProposeUpgrade.Plan,
+	)
+
+	// check proposed upgrade for being deleted
+	_, isFound := setup.Keeper.GetProposedUpgrade(setup.Ctx, msgProposeUpgrade.Plan.Name)
+	require.False(t, isFound)
+}
+
+func TestHandler_ProposedUpgradeDoesNotExist(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := testdata.GenerateAccAddress()
+	trusteeAccAddress2 := testdata.GenerateAccAddress()
+	trusteeAccAddress3 := testdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// approve upgrade
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress1)
+	_, err := setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.Error(t, err)
+	require.True(t, types.ErrProposedUpgradeDoesNotExist.Is(err))
+}
+
+func TestHandler_MessageCreatorAlreadyApprovedUpgrade(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := testdata.GenerateAccAddress()
+	trusteeAccAddress2 := testdata.GenerateAccAddress()
+	trusteeAccAddress3 := testdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// propose new upgrade
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+
+	setup.UpgradeKeeper.On("ScheduleUpgrade", mock.Anything, msgProposeUpgrade.Plan).Return(nil)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+
+	// approve upgrade
+	msgApproveUpgrade := NewMsgApproveUpgrade(trusteeAccAddress1)
+	_, err = setup.Handler(setup.Ctx, msgApproveUpgrade)
+	require.Error(t, err)
+	require.True(t, sdkerrors.ErrUnauthorized.Is(err))
+}
+
+func TestHandler_ProposeUpgradePlanHeightLessBlockHeight(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := testdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(1)
+
+	// propose new upgrade with plan height < block height
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+	msgProposeUpgrade.Plan.Height = 1
+	setup.Ctx = setup.Ctx.WithBlockHeight(100)
+
+	setup.UpgradeKeeper.On("ScheduleUpgrade", mock.Anything, msgProposeUpgrade.Plan).Return(sdkerrors.ErrInvalidRequest)
```

Comment:
Please add a comment why `UpgradeKeeper.ScheduleUpgrade` returns `ErrInvalidRequest` in this case.

Reply:
added